bootstore/src/schemes/v0/state.rs

```diff
@@ -0,0 +1,594 @@
+            RackSecretState::Empty => (),
+        }
+        Output::none()
+    }
+
+    fn validate_share(
+        from: &Baseboard,
+        share: &Vec<u8>,
+        share_digests: &Vec<Sha3_256Digest>,
+    ) -> Result<(), ApiError> {
+        let computed_hash = Sha3_256Digest(
+            Sha3_256::digest(share).as_slice().try_into().unwrap(),
+        );
+
+        if !share_digests.contains(&computed_hash) {
+            Err(ApiError::InvalidShare { from: from.clone() })
+        } else {
+            Ok(())
+        }
+    }
+
+    // Send a request for a share to each peer
+    pub fn broadcast_share_requests(
+        &self,
+        rack_uuid: Uuid,
+        peers: &BTreeSet<Baseboard>,
+    ) -> (Uuid, Output) {
+        // All broadcast requests share the same id
+        let request_id = Uuid::new_v4();
+        let envelopes = peers
+            .iter()
+            .cloned()
+            .map(|to| Envelope {
+                to,
+                msg: Request {
+                    id: request_id,
+                    type_: RequestType::GetShare { rack_uuid },
+                }
+                .into(),
+            })
+            .collect();
+
+        (request_id, Output { persist: false, envelopes, api_output: None })
+    }
+
+    /// We received a share
+    ///
+    /// Verify that the share is valid and recompute the secret if there are
+    /// enough shares.
+    pub fn on_share(
+        &mut self,
+        from: Baseboard,
+        request_id: Uuid,
+        share: Vec<u8>,
+        rack_secret_expiry: Ticks,
+    ) -> Output {
+        match self {
+            RackSecretState::Empty => ApiError::UnexpectedResponse {
+                from,
+                state: "-",
+                request_id,
+                msg: "share",
+            }
+            .into(),
+            RackSecretState::Retrieving {
+                from: from_all,
+                shares,
+                threshold,
+                share_digests,
+                ..
+            } => {
+                // Ignore the share if we already have one from this peer
+                if from_all.contains(&from) {
+                    return Output::none();
+                }
+
+                // Compute and validate hash of the received key share
+                if let Err(api_error) =
+                    Self::validate_share(&from, &share, &share_digests)
+                {
+                    return api_error.into();
+                }
+
+                // Keep track of our new valid share
+                shares.insert(share);
+                from_all.insert(from);
+
+                // If we have enough shares, try to recompute the rack secret
+                // If we fail here, this is a major problem so we report it and
+                // clear our rack secret state, as we won't be able to succeed.
+                // This error should be effectively impossible to ever hit.
+                if shares.0.len() == *threshold {
```

Comment:
Is it possible to get > 1 share, such that we should be doing `>= *threshold` here?

Reply:
This is no longer relevant, but no it is not possible.  We process each message in a single threaded manner so we will hit `==` and bail.